File: circular_maze.php
<?php
// circular_maze.php — PHP версия

class CircularMaze {
    private $rings, $sectors, $algo, $seed, $solve;
    private $maze, $path;

    public function __construct($rings, $sectors, $algo = 'dfs', $seed = null, $solve = false) {
        $this->rings = $rings;
        $this->sectors = $sectors;
        $this->algo = $algo;
        $this->seed = $seed;
        $this->solve = $solve;
        $this->maze = null;
        $this->path = null;
        if ($seed !== null) mt_srand($seed);
    }

    private function getNeighbors($ring, $sector) {
        $neighbors = [];
        if ($ring > 0) $neighbors[] = [$ring-1, $sector, 0];
        if ($ring < $this->rings-1) $neighbors[] = [$ring+1, $sector, 1];
        $neighbors[] = [$ring, ($sector-1+$this->sectors)%$this->sectors, 2];
        $neighbors[] = [$ring, ($sector+1)%$this->sectors, 3];
        return $neighbors;
    }

    private function generateDFS() {
        $maze = array_fill(0, $this->rings, array_fill(0, $this->sectors, ['v' => true, 'h' => true]));
        $stack = [[0, 0]];
        $visited = ['0,0' => true];

        while (!empty($stack)) {
            $pos = end($stack);
            $ring = $pos[0];
            $sector = $pos[1];
            $neighbors = $this->getNeighbors($ring, $sector);
            shuffle($neighbors);

            $found = false;
            foreach ($neighbors as $n) {
                list($nr, $ns, $dir) = $n;
                $key = "$nr,$ns";
                if (isset($visited[$key])) continue;
                $visited[$key] = true;
                if ($dir == 0) $maze[$nr][$ns]['h'] = false;
                else if ($dir == 1) $maze[$ring][$sector]['h'] = false;
                else if ($dir == 2) $maze[$ring][($sector-1+$this->sectors)%$this->sectors]['v'] = false;
                else $maze[$ring][$sector]['v'] = false;
                $stack[] = [$nr, $ns];
                $found = true;
                break;
            }
            if (!$found) array_pop($stack);
        }
        return $maze;
    }

    public function generate() {
        $this->maze = $this->generateDFS();
        if ($this->solve) $this->path = $this->solveMaze();
        return $this->maze;
    }

    private function solveMaze() {
        $stack = [[[0, 0]]];
        $visited = ['0,0' => true];

        while (!empty($stack)) {
            $path = array_pop($stack);
            $last = end($path);
            list($ring, $sector) = $last;
            if ($ring == $this->rings-1) return $path;

            foreach ($this->getNeighbors($ring, $sector) as $n) {
                list($nr, $ns, $dir) = $n;
                $key = "$nr,$ns";
                if (isset($visited[$key])) continue;
                $hasWall = false;
                if ($dir == 0) $hasWall = $this->maze[$nr][$ns]['h'];
                else if ($dir == 1) $hasWall = $this->maze[$ring][$sector]['h'];
                else if ($dir == 2) $hasWall = $this->maze[$ring][($sector-1+$this->sectors)%$this->sectors]['v'];
                else $hasWall = $this->maze[$ring][$sector]['v'];
                if (!$hasWall) {
                    $visited[$key] = true;
                    $newPath = $path;
                    $newPath[] = [$nr, $ns];
                    $stack[] = $newPath;
                }
            }
        }
        return null;
    }

    public function printASCII() {
        echo "\033[36m\nКарта лабиринта (вид сверху):\033[0m\n";
        echo "\033[33m  Колец: {$this->rings}, Секторов: {$this->sectors}\033[0m\n";
        for ($ring = 0; $ring < $this->rings; $ring++) {
            echo "R$ring: ";
            for ($sector = 0; $sector < $this->sectors; $sector++) {
                $v = $this->maze[$ring][$sector]['v'] ? '|' : ' ';
                $h = $this->maze[$ring][$sector]['h'] ? '─' : ' ';
                echo $v . $h;
            }
            echo "\n";
        }
    }
}

function main($argv) {
    $rings = 8;
    $sectors = 12;
    $algo = 'dfs';
    $seed = null;
    $solve = false;

    for ($i = 1; $i < count($argv); $i++) {
        if ($argv[$i] == '--rings') { $rings = (int)$argv[++$i]; }
        else if ($argv[$i] == '--sectors') { $sectors = (int)$argv[++$i]; }
        else if ($argv[$i] == '--algo') { $algo = $argv[++$i]; }
        else if ($argv[$i] == '--seed') { $seed = (int)$argv[++$i]; }
        else if ($argv[$i] == '--solve') { $solve = true; }
    }

    echo "\033[36m⭕ Circular Maze Generator (PHP)\033[0m\n";
    echo "📐 Параметры: {$rings} колец, {$sectors} секторов, алгоритм: {$algo}\n";

    $gen = new CircularMaze($rings, $sectors, $algo, $seed, $solve);
    $gen->generate();
    $gen->printASCII();
}

$argc = $_SERVER['argc'] ?? 0;
$argv = $_SERVER['argv'] ?? [];
main($argv);
?>

File: circular_maze.rb
# circular_maze.rb — Ruby версия

class CircularMaze
  attr_reader :maze, :path

  def initialize(rings, sectors, algo = 'dfs', seed = nil, solve = false)
    @rings = rings
    @sectors = sectors
    @algo = algo
    @seed = seed
    @solve = solve
    @maze = nil
    @path = nil
    @rng = seed ? Random.new(seed) : Random.new
  end

  def get_neighbors(ring, sector)
    neighbors = []
    neighbors << [ring-1, sector, 0] if ring > 0
    neighbors << [ring+1, sector, 1] if ring < @rings - 1
    neighbors << [ring, (sector-1+@sectors)%@sectors, 2]
    neighbors << [ring, (sector+1)%@sectors, 3]
    neighbors
  end

  def generate_dfs
    maze = Array.new(@rings) { Array.new(@sectors) { {v: true, h: true} } }
    stack = [[0, 0]]
    visited = Set.new
    visited.add([0, 0])

    while !stack.empty?
      ring, sector = stack.last
      neighbors = get_neighbors(ring, sector).shuffle(random: @rng)

      found = false
      neighbors.each do |nr, ns, dir|
        next if visited.include?([nr, ns])
        visited.add([nr, ns])
        case dir
        when 0 then maze[nr][ns][:h] = false
        when 1 then maze[ring][sector][:h] = false
        when 2 then maze[ring][(sector-1+@sectors)%@sectors][:v] = false
        when 3 then maze[ring][sector][:v] = false
        end
        stack << [nr, ns]
        found = true
        break
      end
      stack.pop unless found
    end
    maze
  end

  def generate
    @maze = generate_dfs
    @path = solve_maze if @solve
    @maze
  end

  def solve_maze
    stack = [[[0, 0]]]
    visited = Set.new
    visited.add([0, 0])

    while !stack.empty?
      path = stack.pop
      ring, sector = path.last
      return path if ring == @rings - 1

      get_neighbors(ring, sector).each do |nr, ns, dir|
        next if visited.include?([nr, ns])
        has_wall = case dir
        when 0 then @maze[nr][ns][:h]
        when 1 then @maze[ring][sector][:h]
        when 2 then @maze[ring][(sector-1+@sectors)%@sectors][:v]
        when 3 then @maze[ring][sector][:v]
        end
        unless has_wall
          visited.add([nr, ns])
          new_path = path.dup << [nr, ns]
          stack << new_path
        end
      end
    end
    nil
  end

  def print_ascii
    puts "\e[36m\nКарта лабиринта (вид сверху):\e[0m"
    puts "\e[33m  Колец: #{@rings}, Секторов: #{@sectors}\e[0m"
    @rings.times do |ring|
      print "R#{ring}: "
      @sectors.times do |sector|
        v = @maze[ring][sector][:v] ? '|' : ' '
        h = @maze[ring][sector][:h] ? '─' : ' '
        print v + h
      end
      puts
    end
  end
end

def main
  rings = 8
  sectors = 12
  algo = 'dfs'
  seed = nil
  solve = false

  args = ARGV
  i = 0
  while i < args.size
    case args[i]
    when '--rings' then rings = args[i+1].to_i; i += 2
    when '--sectors' then sectors = args[i+1].to_i; i += 2
    when '--algo' then algo = args[i+1]; i += 2
    when '--seed' then seed = args[i+1].to_i; i += 2
    when '--solve' then solve = true; i += 1
    else i += 1
    end
  end

  puts "\e[36m⭕ Circular Maze Generator (Ruby)\e[0m"
  puts "📐 Параметры: #{rings} колец, #{sectors} секторов, алгоритм: #{algo}"

  gen = CircularMaze.new(rings, sectors, algo, seed, solve)
  gen.generate
  gen.print_ascii
end

main if __FILE__ == $0
